# Especialização em Ciência de Dados - PUC-Rio
# Machine Learning
## Exercício 3: Problemas de Regressão



---


__Alunos:__ <br>
        Cícero Felipe Ricci Bressane <br>
        José Douglas Nascimento <br>
        Ronaldo Costa <br>
        Victor Mouffron Carvalho Machado

---



### Escolha um dataset simples para problemas de classificação diferente do usado em aula (preferencialmente do UCI Machine Learning Repository - https://archive.ics.uci.edu/ml/index.php) e treine os modelos de regressão aprendidos em aula (Regressão Linear, Ridge, Lasso, ElasticNet, Árvore de Regressão, KNN, SVR).

O dataset escolhido foi obtido no UCI no link https://archive.ics.uci.edu/ml/datasets/Wine+Quality. Trata-se de um dataset com características físico-químicas e sensorias de vinhos. O objetivo é prever a qualidade do vinho avaliada de 0 a 10. Optou-se por utilizar apenas o dataset relativo à vinhos brancos.


__Vamos aplicar os modelos!__


![Vinho](img/wine.jpg)

In [38]:
import warnings
import pandas as pd
warnings.filterwarnings('ignore')

df = pd.read_csv('data/winequality-white.csv',sep=';')
df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  quality  
0      8.8        6  
1      9.5        6  
2     10.1        6  
3      9.9        6  
4      9.9        6

In [39]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4898 entries, 0 to 4897
Data columns (total 12 columns):
fixed acidity           4898 non-null float64
volatile acidity        4898 non-null float64
citric acid             4898 non-null float64
residual sugar          4898 non-null float64
chlorides               4898 non-null float64
free sulfur dioxide     4898 non-null float64
total sulfur dioxide    4898 non-null float64
density                 4898 non-null float64
pH                      4898 non-null float64
sulphates               4898 non-null float64
alcohol                 4898 non-null float64
quality                 4898 non-null int64
dtypes: float64(11), int64(1)
memory usage: 459.3 KB


Temos 4898 instâncias e nenhum dado faltante. Vamos separar o dataset em treino e teste

In [40]:
from sklearn.model_selection import train_test_split
random_seed = 0
X = df.iloc[:,:-1].values
y = df.iloc[:,-1].values
X_train, X_test, y_train, y_test = train_test_split(X,y,random_state=random_seed)

In [41]:
from sklearn.preprocessing import RobustScaler

print('Fazendo scaling dos datasets... ')
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

Fazendo scaling dos datasets: 


In [44]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso
from sklearn.linear_model import ElasticNet
from sklearn.svm import SVR
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import mean_squared_error
import numpy as np

models = [
            ('DecisionTreeRegressor',DecisionTreeRegressor(random_state=random_seed)),
            ('LinearRegression',LinearRegression()),
            ('Ridge',Ridge()),
            ('Lasso',Lasso()),
            ('SVR',SVR()),
            ('ElasticNet',ElasticNet(random_state=random_seed)),
            ('KNeighborsRegressor',KNeighborsRegressor())
          ]

for name,model in models:
    model.fit(X_train_scaled,y_train)
    y_pred = model.predict(X_test_scaled)
    rmse = np.sqrt(mean_squared_error(y_pred,y_test)) 
    print('O erro do modelo ',name,' foi de ',rmse)

O erro do modelo  DecisionTreeRegressor  foi de  0.8699049912936188
O erro do modelo  LinearRegression  foi de  0.7955515218421332
O erro do modelo  Ridge  foi de  0.7953008890117879
O erro do modelo  Lasso  foi de  0.9176196407505608
O erro do modelo  SVR  foi de  0.7316094899184977
O erro do modelo  ElasticNet  foi de  0.9176196407505608
O erro do modelo  KNeighborsRegressor  foi de  0.7551293982909586


### Agora pesquise outras bibliotecas, modelos de regressão ou funcionalidades (não estudados em sala) e aplique-os no problema. Lembre-se que quanto mais você pesquisar e treinar, mais irá aprender.**

Além desses modelos temos os ensembles, vistos na Aula 5. Uma biblioteca que tem tido muito sucesso em competições de ML é a xgboost que implementa um ensemble. Para mais informações ver o paper Chen & Guestrin (2016) que descreve esse algoritmo, disponível em https://arxiv.org/pdf/1603.02754.pdf - XGBoost: A Scalable Tree Boosting System.

In [45]:
import xgboost as xgb
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.ensemble import AdaBoostRegressor


models = [
            ('XGBoost',xgb.XGBRegressor(random_state=random_seed,objective ='reg:squarederror')),
            ('Random forest',RandomForestRegressor(random_state=random_seed)),
            ('Gradient Boosting',GradientBoostingRegressor(random_state=random_seed)),
            ('Ada Boosting',AdaBoostRegressor(random_state=random_seed))
          ]

for name,model in models:
    model.fit(X_train_scaled,y_train)
    y_pred = model.predict(X_test_scaled)
    rmse = np.sqrt(mean_squared_error(y_pred,y_test)) 
    print('O erro do modelo ',name,' foi de ',rmse)

O erro do modelo  XGBoost  foi de  0.7396279090786261
O erro do modelo  Random forest  foi de  0.6961292397014701
O erro do modelo  Gradient Boosting  foi de  0.7341928665522126
O erro do modelo  Ada Boosting  foi de  0.787355129670329


### O que você aprendeu sobre o problema de regressão deste dataset?

Os resultados sugeriram um bom desempenho dos ensembles, sendo que a Random Forest obteve melhor resultado. Além disso, ao executar os modelos de ensemble sem o scaling percebemos que esses modelos praticamente não foram afetados pela mudança na dimensão das variáveis, enquanto o  KNeighborsRegressor teve uma melhora substancial no seu desempenho.